### ISSUE TITLE
could certain schemas derived from enums more precisely use oneOf?

### ISSUE THREAD

**Issue by ahl** [GREsau/schemars#98]
When writing an SDK generator, we noticed that schemas derived from enums use `anyOf` for complex variants (and `enum` for unit variants). This seems necessary in some cases, for example:

```rust
#[derive(JsonSchema)]
#[serde(untagged)]
pub enum E {
    ThingA { a: String },
    ThingB { a: String },
}
```

i.e. where the variants literally cannot be disambiguated. In such a case I don't believe that `oneOf` would be satisfiable since two subschemas would be identical.

However with tagged enums (external, internal, or adjacent) would it be reasonable to change `anyOf` to `oneOf`? For example in this output of the type above, removing the `untagged` directive:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "E",
  "anyOf": [
    {
      "type": "object",
      "required": [
        "ThingA"
      ],
      "properties": {
        "ThingA": {
          "type": "object",
          "required": [
            "a"
          ],
          "properties": {
            "a": {
              "type": "string"
            }
          }
        }
      },
      "additionalProperties": false
    },
    {
      "type": "object",
      "required": [
        "ThingB"
      ],
      "properties": {
        "ThingB": {
          "type": "object",
          "required": [
            "a"
          ],
          "properties": {
            "a": {
              "type": "string"
            }
          }
        }
      },
      "additionalProperties": false
    }
  ]
}
```

In this case, where each subschema is mutually exclusive, I believe that `anyOf` and `oneOf` are equivalent. It would, however, be more explicit to use `oneOf` and, as noted earlier, it would simplify our particular use case of SDK generation.

If this seems like a reasonable change I'd be happy to submit a PR, but I wanted to check in first.

**Comment by jhoobergs** [comment]
What do you think of https://github.com/GREsau/schemars/pull/96 ?

**Comment by GREsau** [comment]
> However with tagged enums (external, internal, or adjacent) would it be reasonable to change anyOf to oneOf?

Yes that sounds reasonable, and I don't *think* it would be difficult to implement. I'll take a look when I get some more time - I'd also accept this as a PR

**Comment by ahl** [comment]
Thanks! I submitted as #108

**Comment by GREsau** [comment]
This behaviour is now available in v0.8.6
